### PR TITLE
Initial Auth Next implementation in Soroban Host

### PR DESCRIPTION
**Pull request by dmkozh**
### What

This is the initial implementation iteration of the Auth Next [proposal](https://docs.google.com/document/d/1J-J3ClTUkrsLiJag906OH4hmNkZI3Jk6_Y9ZYt_psAI/edit#heading=h.mne05sjh63jf).

This is based on the XDR changes (to be merged) from https://github.com/stellar/stellar-xdr/pull/65

On a high level, this PR introduces the following:

- Support for the unified `Address` host object type that can be used to represent either classic Stellar accounts or generic contract-based account. This replaces `AccountId` and `Identifier` types, as well as `invoker`/`source_account` etc. in the public interfaces.  `Address` can be passed around via contract interfaces, used in the data storage keys, as well as used in authorization flows.
- Implementation of the `require_auth` host function that enforces an authorization of the current contract call from the given `Address`. The authorization data with the corresponding signature(s) is passed along with the contract invocation transaction.
- Behind the scenes of `require_auth` is `AuthorizationManager` (`auth.rs`) that keeps track of the call stacks to provide the proper signature context, ensures that authorizations aren't reused, and performs authentication using the proper account contract. It also has a recording mode that allows to record `require_auth` calls with the context instead of enforcing them.
- Updated of the built-in token contract to use the Auth Next functionality
- Basic support for the custom account contracts

### Why

The Auth Next motivation is in the proposal. The general idea is that we want to provide a protocol-defined support for generalized authorization of nested contract invocations, customizable authentication via account abstraction and structured signatures payloads that allow for better wallet integration.

### Known limitations

This does some foundational work on Auth Next, and followups of various scope are required. Some notable followups are:
- More test coverage, especially for deep/wide contract invocation trees.
- Better error reporting with corresponding new error types
- Better support for custom account contracts (protect `check_auth` to allow for safe ledger writes, allow only self-reentrance, maybe update interface etc.)
- Temporary/ephemeral storage implementation for safer implementation of allowance based patterns and corresponding token update
- Temporary/ephemeral nonces (instead of auto-increment) for better concurrency/lower ledger space usage/easier signature flows

### CODE REVIEW

**Comment by dmkozh** [comment]
Build is supposed to be failing until the XDR changes are merged and pinned here.

**Review by sisuresh — COMMENTED** [review]
Still reviewing. Just a couple comments at the moment.

File: soroban-env-host/src/auth.rs
Comment (dmkozh) [reply soroban-env-host/src/auth.rs:359]: Nice catch, I've actually noticed this only when ran some tests. This should be fixed now (as well as another recording-related bug)

File: soroban-env-host/src/host/metered_clone.rs
Comment (dmkozh) [reply soroban-env-host/src/host/metered_clone.rs]: That's actually unnecessary, just an artifact from the old iteration. Removed.

File: soroban-env-host/src/auth.rs
Comment (dmkozh) [reply soroban-env-host/src/auth.rs]: Fixed.

File: soroban-env-host/src/auth.rs
Comment (dmkozh) [reply soroban-env-host/src/auth.rs:359]: To be clear, the idea is to completely reset the whole `AuthorizationManager`. The only reason for not creating a new instance is to store the previous authorizations in order to be able to verify them. Maybe it would actually make more sense to do the backup one level higher (i.e. in the host). Does that sound reasonable?

File: soroban-env-host/src/host.rs
Comment (sisuresh) [inline soroban-env-host/src/host.rs:2289]: Using the handle to track different `require_auth` calls for the same `ScAddress` in recording mode looks like a footgun to me. If I understood @graydon at standup correctly, we are considering making these objects unique (so handles will be the same). A change like that would break auth. This shouldn't hold up the release, but it's something we should discuss for the next auth update.

File: soroban-env-host/src/auth.rs
Comment (sisuresh) [reply soroban-env-host/src/auth.rs:359]: Is `testutils` required for recording mode? I see you added `tracker_by_address_handle.clear()` in `AuthorizationManager::reset()`, but that's behind a `testutils` conditional.

And I don't think I understand your question (I think my misunderstanding here is related to the above point). The host already calls `AuthorizationManager::reset()` right?

Comment (leighmcculloch) [reply soroban-env-host/src/auth.rs:359]: > Is `testutils` required for recording mode?

Recording mode will be used by non-tests and needs to exist outside the testutils feature. For example the soroban-rpc imports the soroban-env-host and will need to use the recording mode.

Comment (dmkozh) [reply soroban-env-host/src/auth.rs:359]: > Is testutils required for recording mode? I see you added tracker_by_address_handle.clear() in AuthorizationManager::reset(), but that's behind a testutils conditional.

The whole `reset` method is test-specific, hence it's only enabled when `testutils` feature is present. The common uint testing pattern is to reuse the same host instance for multiple contract invocations, which is why we need to reset auth for every invocation in order to emulate the real behavior. In not-test mode both enforcing and recording auth shouldn't be reset - instead different instance of host would be needed. This may or may not need an adjustment when we allow multi-op transactions, but for now this reflects the current state of the things.

File: soroban-env-host/src/host.rs
Comment (dmkozh) [reply soroban-env-host/src/host.rs:2289]: > If I understood @graydon at standup correctly, we are considering making these objects unique (so handles will be the same).

I'm not sure I understood, would we somehow have a single set of objects and do the lookups? That doesn't sound efficient...

>  A change like that would break auth. 
FWIW this is used in the recording mode only. But if we somehow would try to deduplicate objects, then `Address` host object type would need an additional level of indirection in order to distinguish between different objects.

File: soroban-env-host/src/auth.rs
Comment (sisuresh) [reply soroban-env-host/src/auth.rs:359]: So I think I got a bit off track here - you mentioned that you fixed the bug at the top of this thread by resetting `AuthorizationManager`. Where is that fix exactly?

Comment (dmkozh) [reply soroban-env-host/src/auth.rs:359]: The fix is in this commit: https://github.com/stellar/rs-soroban-env/pull/645/commits/4a4a8d43b1e7fa941e932041d6a42f6f28b466cf

What I'm thinking about is to remove the `reset` method completely and instead keep a backup of `AuthorizationManager` in tests.

Comment (sisuresh) [reply soroban-env-host/src/auth.rs:359]: So you're saying this bug can only happen in tests?

Comment (dmkozh) [reply soroban-env-host/src/auth.rs:359]: Yeah, in non-test environment there is nothing to reset in the first place.

Comment (sisuresh) [reply soroban-env-host/src/auth.rs:359]: As we discussed offline, `tracker_by_address_handle` needs to be part of the auth managers snapshot.

**Review by sisuresh — APPROVED** [review]

File: soroban-env-host/src/auth.rs
Comment (sisuresh) [inline soroban-env-host/src/auth.rs:212]: Why is this an `Option`?

Comment (dmkozh) [reply soroban-env-host/src/auth.rs:212]: The task of the tracker is to map the 'condensed' tree of authorizations to the current call stack. This function returns the bottom-most matching authorization tree node. Since the current call stack may not have any matches for this particular tracker there won't be any node to return here, hence this is an option.

FWIW in the recording mode we lazily create trackers on first invocation (thus making it the tree root), so this should never return none there. But for the enforcing mode it's perfectly valid to return `None` here as the trackers are created based on the transaction field.

File: soroban-env-host/src/auth.rs
Comment (sisuresh) [inline soroban-env-host/src/auth.rs:754]: Are there any weird edge cases that can result from having duplicate sub_invocations? The frame index would be overwritten here and a sub_invocation will become exhausted even if it isn't used. So a contract that calls `authorize` from the same account with the same args twice would never work right? I'm guessing we should break here.

File: soroban-env-host/src/auth.rs
Comment (sisuresh) [reply soroban-env-host/src/auth.rs:212]: But this function only ever returns `Some(self)` right? I think you're referring to `AuthorizationTracker::last_authorized_invocation_mut` and not `AuthorizedInvocation::last_authorized_invocation_mut` (where I made this comment).

Comment (dmkozh) [reply soroban-env-host/src/auth.rs:212]: Sorry, you're right. That's just a refactoring artifact.

File: soroban-env-host/src/auth.rs
Comment (dmkozh) [reply soroban-env-host/src/auth.rs:754]: Nice catch, there is really no reason not to break here; that's just a bug.

**Pull request merged**
